### PR TITLE
add exceptions for tokens function.

### PR DESCRIPTION
**Pull request by ObiWahn**

### CODE REVIEW

**Comment by jsteemann** [comment]
Most of the changes seem to be unrelated whitespace changes.
Can we try to keep PRs as minimal as possible? Otherwise they are really hard to review.

**Comment by ObiWahn** [comment]
I talked with @gnusi about that. I wanted to clang format some longer line and the result was that. The problem is that the code is not formatted as it should be. @gnusi wants to reformat the code as soon as most of the bugs are fixed.

Furthermore we agreed on clang-formatting and I do not see a problem on my side. If everybody would follow the rules set up by @fceller the diff would be concise.

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/4442/

**Comment by jsteemann** [comment]
We also agreed on keeping PRs minimal.
I will not review PRs that have thousands of lines with unnecessary changes.
The PR already has merge conflicts now.

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/4475/

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 1]
```diff
@@ -135,27 +137,24 @@ bool IdentityAnalyzer::reset(irs::string_ref const& data) {
 arangodb::aql::AqlValue aqlFnTokens(arangodb::aql::ExpressionContext* expressionContext,
                                     arangodb::transaction::Methods* trx,
                                     arangodb::aql::VPackFunctionParameters const& args) {
-  if (2 != args.size() || !args[0].isString() || !args[1].isString()) {
-    LOG_TOPIC("740fd", WARN, arangodb::iresearch::TOPIC)
-        << "invalid arguments passed while computing result for function "
-           "'TOKENS'";
-    TRI_set_errno(TRI_ERROR_BAD_PARAMETER);
+  std::string const while_tokens =
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: let's avoid allocating strings for good case

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 2]
```diff
@@ -135,27 +137,24 @@ bool IdentityAnalyzer::reset(irs::string_ref const& data) {
...
-    return arangodb::aql::AqlValue();
+  if (2 != args.size() || !args[0].isString() || !args[1].isString()) {
+    auto message = "invalid arguments" + while_tokens;
+    LOG_TOPIC("740fd", WARN, arangodb::iresearch::TOPIC) << message;
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_BAD_PARAMETER, message);
   }
 
   auto data = arangodb::iresearch::getStringRef(args[0].slice());
-  auto name = arangodb::iresearch::getStringRef(args[1].slice());
+  auto name = args[1].slice().copyString();
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: let's avoid allocating strings for good case

**Comment by ObiWahn** [comment]
The string ref is not zero terminated and lacks operator+ for strings. I guess that needs to be fixed upstream. I thought about adding it but I did not want that you get out of sync or need to port from Arango to IResearch. Changing to string seems to me to be the least intrusive change to not read outside your allocation. Think the speedup does not justify to have several conversions in the code.

File: arangod/IResearch/IResearchAnalyzerFeature.cpp (hunk 1 above)
Comment (ObiWahn) [reply arangod/IResearch/IResearchAnalyzerFeature.cpp]: I can pull it out the function and make it static.

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/4482/

**Comment by gnusi** [comment]
@ObiWahn there are no overloads since `string_ref` in iresearch can be explicitly converted to `std::string`

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 3]
```diff
@@ -132,16 +134,26 @@ bool IdentityAnalyzer::reset(irs::string_ref const& data) {
   return !_empty;
 }
 
+static std::string const while_tokens =
+      " while computing result for function 'TOKENS'";
+
+namespace detail {
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]:
```suggestion
```

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 4]
```diff
@@ -177,34 +178,38 @@ arangodb::aql::AqlValue aqlFnTokens(arangodb::aql::ExpressionContext* expression
   }
 
   if (!pool) {
+    auto n = std::string(name.c_str(), name.size());
     auto const message = "failure to find arangosearch analyzer with name '"s +
-                         name + "'" + while_tokens;
+                         n + "'" + while_tokens;
```
Comment (graetzer) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp]: how about `std::string("failure to find ...").append(name.c_str(), name.size()).append("'")` etc

**Pull request merged**